packages/flutter/lib/src/widgets/linked_text.dart

```diff
@@ -0,0 +1,614 @@
+
+  /// Returns a [LinkBuilder] that highlights the given text and sets the given
+  /// [onTap] handler.
+  static LinkBuilder getDefaultLinkBuilder(LinkTapCallback onTap) {
+    return (String displayString, String linkString) {
+      final TapGestureRecognizer recognizer = TapGestureRecognizer()
+          ..onTap = () => onTap(linkString);
+      return (
+        InlineLink(
+          recognizer: recognizer,
+          text: displayString,
+        ),
+        recognizer,
+      );
+    };
+  }
+
+  /// The spans on which to create links by applying [textLinkers].
+  final List<InlineSpan> spans;
+
+  /// Defines what parts of the text to match and how to link them.
+  ///
+  /// [TextLinker]s are applied in the order given. Overlapping matches are not
+  /// supported.
+  late final List<TextLinker> textLinkers;
+
+  /// The [TextStyle] to apply to the output [InlineSpan].
+  final TextStyle? style;
+
+  /// Apply the given [TextLinker]s to the given [InlineSpan]s and return the
+  /// new resulting spans and any created [TapGestureRecognizer]s.
+  ///
+  /// {@macro flutter.painting.LinkBuilder.recognizer}
+  static (Iterable<InlineSpan>, Iterable<TapGestureRecognizer>) linkSpans(Iterable<InlineSpan> spans, Iterable<TextLinker> textLinkers) {
+    // Flatten the spans and find all ranges in the flat String. This must be done
+    // cumulatively, and not during a traversal, because matches may occur across
+    // span boundaries.
+    final List<({InlineSpan span, int length})> spansWithLength =
+        <({InlineSpan span, int length})>[];
+    String spansText = '';
+    for (final InlineSpan span in spans) {
+      final String string = span.toPlainText();
+      spansText += string;
+      spansWithLength.add((
+        span: span,
+        length: string.length,
+      ));
+    }
+    final Iterable<_TextLinkerMatch> textLinkerMatches =
+        _cleanTextLinkerMatches(
+          _TextLinkerMatch.fromTextLinkers(textLinkers, spansText),
+        );
+
+    final (Iterable<InlineSpan> output, Iterable<_TextLinkerMatch> _, Iterable<TapGestureRecognizer> recognizers) =
+        _linkSpansRecurse(spansWithLength, textLinkerMatches, 0);
+    return (output, recognizers);
+  }
+
+  static List<_TextLinkerMatch> _cleanTextLinkerMatches(Iterable<_TextLinkerMatch> textLinkerMatches) {
+    final List<_TextLinkerMatch> nextTextLinkerMatches = textLinkerMatches.toList();
+
+    // Sort by start.
+    nextTextLinkerMatches.sort((_TextLinkerMatch a, _TextLinkerMatch b) {
+      return a.textRange.start.compareTo(b.textRange.start);
+    });
+
+    int lastEnd = 0;
+    nextTextLinkerMatches.removeWhere((_TextLinkerMatch textLinkerMatch) {
+      // Return empty ranges.
+      if (textLinkerMatch.textRange.start == textLinkerMatch.textRange.end) {
+        return true;
+      }
+
+      // Remove overlapping ranges.
+      final bool overlaps = textLinkerMatch.textRange.start < lastEnd;
```

Comment:
We could also use textlinker order to determine which should be throw away. Right now it seems the behavior is hard to predict?

Reply:
I changed it to throw when an overlap is found.

I looked at the behavior of [RegExp.allMatches](https://master-api.flutter.dev/flutter/dart-core/RegExp/allMatches.html), and it won't produce overlapping ranges by default.  So if a user does manage to produce overlapping matches, they are probably doing something strange, and it's probably best to notify them rather than to silently remove them.